src/analysis/processing/qgsalgorithmlinedensity.cpp

```diff
@@ -0,0 +1,193 @@
+{
+  return new QgsLineDensityAlgorithm();
+}
+
+bool QgsLineDensityAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  mSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+  mWeight = parameterAsString( parameters, QStringLiteral( "WEIGHT" ), context );
+  mSearchRadius = parameterAsDouble( parameters, QStringLiteral( "RADIUS" ), context );
+  mPixelSize = parameterAsDouble( parameters, QStringLiteral( "PIXEL_SIZE" ), context );
+
+  mExtent = mSource->sourceExtent();
+  mCrs = mSource->sourceCrs();
+
+  if ( context.project() )
+  {
+    mDa.setEllipsoid( context.project()->ellipsoid() );
+  }
+  mDa.setSourceCrs( mCrs, context.transformContext() );
+
+
+  //get cell midpoint from top left cell
+  QgsGeometry firstCellMidpoint = QgsGeometry( new QgsPoint( mExtent.xMinimum() + ( mPixelSize / 2 ), mExtent.yMaximum() - ( mPixelSize / 2 ) ) );
+  mSearchGeometry = firstCellMidpoint.buffer( mSearchRadius, 5 );
+  mSearchGeometryArea = mDa.measureArea( mSearchGeometry );
+
+  mIndex = QgsSpatialIndex( *mSource, nullptr, QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+
+  return true;
+}
+
+QVariantMap QgsLineDensityAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  int rows = std::max( std::ceil( mExtent.height() / mPixelSize ), 1.0 );
+  int cols = std::max( std::ceil( mExtent.width() / mPixelSize ), 1.0 );
+
+  //build new raster extent based on number of columns and cellsize
+  //this prevents output cellsize being calculated too small
+  QgsRectangle rasterExtent = QgsRectangle( mExtent.xMinimum(), mExtent.yMaximum() - ( rows * mPixelSize ), mExtent.xMinimum() + ( cols * mPixelSize ), mExtent.yMaximum() );
+
+  std::unique_ptr< QgsRasterFileWriter > writer = qgis::make_unique< QgsRasterFileWriter >( outputFile );
+  writer->setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer->createOneBandRaster( Qgis::Float32, cols, rows, rasterExtent, mCrs ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+
+  provider->setNoDataValue( 1, -9999 );
+
+  int totalCellcnt = rows * cols;
```

Comment:
better go with qgssize here instead, or there's a risk of overflows:
```suggestion
  qgssize totalCellcnt = static_cast< qgssize >( rows ) * cols;
```